### PR TITLE
More safe, but still ugly mutations interpreter

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (up to few sentences, not needed for non-significant PRs):
Fix bug in mutations that have predicate with actions that require context (for example functions for json).

### CODE REVIEW

**Comment by alesapin** [comment]
` Merge with master — Cannot fetch mergecommit` -- strange, pull request is mergeable...

File: dbms/src/Interpreters/MutationsInterpreter.cpp [hunk 1]
```diff
@@ -89,10 +89,46 @@ std::optional<String> findFirstNonDeterministicFuncName(const MutationCommand &
 
     return {};
 }
-};
 
+ASTPtr prepareQueryAffectedAST(const std::vector<MutationCommand> & commands)
+{
+    /// Execute `SELECT count() FROM storage WHERE predicate1 OR predicate2 OR ...` query.
+    /// The result can differ from tne number of affected rows (e.g. if there is an UPDATE command that
+    /// changes how many rows satisfy the predicates of the subsequent commands).
```
Comment (KochetovNicolai) [inline dbms/src/Interpreters/MutationsInterpreter.cpp]: Is difficult to understand. Maybe add an example?
If I am not mistaken, the idea is simple: not all the rows we read are satisfy the predicates.

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 2]
```diff
@@ -689,7 +689,7 @@ bool StorageMergeTree::tryMutatePart()
             size_t current_ast_elements = 0;
             for (auto it = mutations_begin_it; it != mutations_end_it; ++it)
             {
-                MutationsInterpreter interpreter(shared_from_this(), it->second.commands, global_context);
+                MutationsInterpreter interpreter(shared_from_this(), it->second.commands, global_context, false);
```
Comment (KochetovNicolai) [inline dbms/src/Storages/StorageMergeTree.cpp]: Why can_execute is false here?

**Review by KochetovNicolai — APPROVED** [review]

**Comment by alesapin** [comment]
Conflicts in 19.15 :(

**Pull request merged**